src/box/decimal.c

```diff
@@ -0,0 +1,220 @@
+}
+
+box_decimal_t *
+box_decimal_from_double(box_decimal_t *dec, double d)
+{
+	return (box_decimal_t *)decimal_from_double((decimal_t *)dec, d);
+}
+
+box_decimal_t *
+box_decimal_from_int64(box_decimal_t *dec, int64_t num)
+{
+	return (box_decimal_t *)decimal_from_int64((decimal_t *)dec, num);
+}
+
+box_decimal_t *
+box_decimal_from_uint64(box_decimal_t *dec, uint64_t num)
+{
+	return (box_decimal_t *)decimal_from_uint64((decimal_t *)dec, num);
+}
+
+box_decimal_t *
+box_decimal_copy(box_decimal_t *dest, const box_decimal_t *src)
+{
+	/*
+	 * Copy only known part of the value. The rest of the
+	 * storage may be unallocated (if we got it from
+	 * tarantool).
+	 */
+	memcpy(dest, src, sizeof(decimal_t));
+	return dest;
+}
+
+/* }}} decimal constructors */
+
+/* {{{ decimal conversions */
+
+void
+box_decimal_to_string(const box_decimal_t *dec, char *buf)
+{
+	decimal_to_string((const decimal_t *)dec, buf);
+}
+
+const box_decimal_t *
+box_decimal_to_int64(const box_decimal_t *dec, int64_t *num)
+{
+	return (const box_decimal_t *)decimal_to_int64(
+		(const decimal_t *)dec, num);
+}
+
+const box_decimal_t *
+box_decimal_to_uint64(const box_decimal_t *dec, uint64_t *num)
+{
+	return (const box_decimal_t *)decimal_to_uint64(
+		(const decimal_t *)dec, num);
+}
+
+/* }}} decimal conversions */
+
+/* {{{ decimal rounding */
+
+box_decimal_t *
+box_decimal_round(box_decimal_t *dec, int scale)
+{
+	return (box_decimal_t *)decimal_round((decimal_t *)dec, scale);
+}
+
+box_decimal_t *
+box_decimal_floor(box_decimal_t *dec, int scale)
+{
+	return (box_decimal_t *)decimal_floor((decimal_t *)dec, scale);
+}
+
+box_decimal_t *
+box_decimal_trim(box_decimal_t *dec)
+{
+	return (box_decimal_t *)decimal_trim((decimal_t *)dec);
+}
+
+box_decimal_t *
+box_decimal_rescale(box_decimal_t *dec, int scale)
+{
+	return (box_decimal_t *)decimal_rescale((decimal_t *)dec, scale);
+}
+
+/* }}} decimal rounding */
+
+/* {{{ decimal arithmetic */
+
+int
+box_decimal_compare(const box_decimal_t *lhs, const box_decimal_t *rhs)
+{
+	return decimal_compare((const decimal_t *)lhs, (const decimal_t *)rhs);
+}
+
+BOX_DECIMAL_UNARY_OP(abs)
+BOX_DECIMAL_UNARY_OP(minus)
+BOX_DECIMAL_BINARY_OP(add)
+BOX_DECIMAL_BINARY_OP(sub)
+BOX_DECIMAL_BINARY_OP(mul)
+BOX_DECIMAL_BINARY_OP(div)
+BOX_DECIMAL_BINARY_OP(remainder)
+
+/* }}} decimal arithmetic */
+
+/* {{{ decimal math functions */
+
+BOX_DECIMAL_UNARY_OP(log10)
+BOX_DECIMAL_UNARY_OP(ln)
+BOX_DECIMAL_BINARY_OP(pow)
+BOX_DECIMAL_UNARY_OP(exp)
+BOX_DECIMAL_UNARY_OP(sqrt)
+
+/* }}} decimal math functions */
+
+/* {{{ decimal encoding to/decoding from msgpack */
+
+uint32_t
+box_decimal_mp_sizeof(const box_decimal_t *dec)
+{
+	return mp_sizeof_decimal((const decimal_t *)dec);
+}
+
+char *
+box_decimal_mp_encode(const box_decimal_t *dec, char *data)
+{
+	return mp_encode_decimal(data, (const decimal_t *)dec);
+}
+
+box_decimal_t *
+box_decimal_mp_decode(box_decimal_t *dec, const char **data)
```

Comment:
There's a separate `decimal_unpack`, because when you're decoding a msgpack string it's generally unknown which extension type it holds. you first do a `mp_decode_extl`, which tells you the extension type, and then you do `type_unpack`, which decodes the remainder.

Otherwise the user would have to do `mp_decode_extl`, restore the buffer position to the beginning of MP_EXT and call `mp_decode_decimal`. This is a little bit extraneous compared to `mp_decode_extl` + `decimal_unpack`.

So I propose to expose `decimal_unpack`.

Reply:
Mostly to remember it for myself, how a value of an extension type is encoded in [msgpack](https://github.com/msgpack/msgpack/blob/master/spec.md) and what functions we have around (including one proposed in this PR):

| msgpack type | size       | extension type | data              | next values |
| ------------ | ---------- | -------------- | ----------------- | ----------- |
| 1 byte       | 0--4 bytes | 1 byte         | 0--(2^32-1) bytes |             |
| `mp_typeof()` is called on this position | | | |
| `mp_decode_extl()` is called on this position | | | `mp_decode_extl()` rewinds the data pointer to here |
| `mp_decode_ext()` is called on this position | | | | `mp_decode_ext()` rewinds the data pointer to here |
| `box_<foo>_mp_decode()` is called from here | | | | `box_<foo>_mp_decode()` rewinds the data pointer to here |

I think that a function with the `box_<foo>_mp_decode()` name should accept a start of the `<foo>` value in an msgpack buffer. However I agree that we should provide a function to decode `<foo>` from buffer that starts from `data`.

I don't like 'unpack' word in this context: it has no relation to msgpack and we use encode/decode terms around msgpack. I would name it `box_<foo>_mp_decode_data()` and accept the data size (which is usually got from `mp_decode_extl()`).

So, we'll add the following line to the table above:

| msgpack type | size       | extension type | data              | next values |
| ------------ | ---------- | -------------- | ----------------- | ----------- |
| | | | `box_<foo>_mp_decode_data()` is called from here | `box_<foo>_mp_decode_data()` rewinds the data pointer to here |

If there are no objections, I'll implement it in this PR.

Reply:
Added `box_decimal_mp_decode_data()`.

Reply:
Sounds good